crates/oci-distribution/src/client.rs

```diff
@@ -114,6 +127,73 @@ impl Client {
         })
     }
 
+    /// Push an image and return the uploaded URL of the image
+    ///
+    /// The client will check if it's already been authenticated and if
+    /// not will attempt to do.
+    ///
+    /// If a manifest is not provided, the client will attempt to generate
+    /// it from the provided image and config data.
+    ///
+    /// Returns pullable URL for the image
+    pub async fn push(
+        &mut self,
+        image_ref: &Reference,
+        image_data: &ImageData,
+        config_data: &[u8],
+        config_media_type: &str,
+        auth: &RegistryAuth,
+        manifest: Option<OciManifest>,
+    ) -> anyhow::Result<String> {
+        debug!("Pushing image: {:?}", image_ref);
+
+        if !self.tokens.contains_key(image_ref.registry()) {
+            self.auth(image_ref, auth, &RegistryOperation::Push).await?;
+        }
+
+        // Start push session
+        let mut location = self.begin_push_session(image_ref).await?;
+
+        // Upload layers
+        let mut start_byte = 0;
+        for layer in &image_data.layers {
+            // Destructuring assignment is not yet supported
+            let (next_location, next_byte) = self
+                .push_layer(&location, &image_ref, layer.data.to_vec(), start_byte)
+                .await?;
+            location = next_location;
+            start_byte = next_byte;
+        }
+
+        // End push session, upload manifest
+        let image_url = self
+            .end_push_session(
+                &location,
+                &image_ref,
+                &sha256_digest(
+                    &image_data
+                        .layers
+                        .clone()
+                        .into_iter()
+                        .map(|l| l.data)
+                        .flatten()
+                        .collect::<Vec<u8>>(),
+                ), // Not ideal copying entire image to compute hash
```

Comment:
Hm, I've been poking around and it seems that Vec<T> doesn't implement the copy trait since it's unsafe to do so (something something doubling buffer size, leading to double free-ing the value later on, https://doc.rust-lang.org/std/marker/trait.Copy.html#when-cant-my-type-be-copy) I think we're forced into cloning the data to get it out of the shared reference. I believe it could be done without cloning if we're willing to drop the value when we compute the hash

Reply:
I could definitely be missing something about iterators, and if I am please let me know! I'd love to make this more efficient, and I dislike how inefficient this is for memory right now 😢